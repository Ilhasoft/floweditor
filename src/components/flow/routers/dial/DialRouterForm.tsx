import { react as bindCallbacks } from 'auto-bind';
import * as React from 'react';
import Dialog, { ButtonSet, Tab } from 'components/dialog/Dialog';
import { hasErrors, renderIssues } from 'components/flow/actions/helpers';
import { RouterFormProps } from 'components/flow/props';
import { nodeToState, stateToNode } from 'components/flow/routers/dial/helpers';
import { createResultNameInput } from 'components/flow/routers/widgets';
import TextInputElement from 'components/form/textinput/TextInputElement';
import TypeList from 'components/nodeeditor/TypeList';
import { FormState, StringEntry } from 'store/nodeEditor';
import { Alphanumeric, Required, StartIsNonNumeric, validate } from 'store/validators';
import i18n from 'config/i18n';
import CheckboxElement from 'components/form/checkbox/CheckboxElement';

export interface DialRouterFormState extends FormState {
  phone: StringEntry;
  resultName: StringEntry;
}

export default class DialRouterForm extends React.Component<RouterFormProps, DialRouterFormState> {
  constructor(props: RouterFormProps) {
    super(props);

    this.state = nodeToState(this.props.nodeSettings);

    bindCallbacks(this, {
      include: [/^on/, /^handle/]
    });
  }

  private handlePhoneUpdated(value: string): void {
    const phone = validate(i18n.t('forms.phone_number', 'Phone Number'), value, [Required]);
    this.setState({
      phone: phone,
      valid: this.state.valid && !hasErrors(phone)
    });
  }

  private handleUpdateResultName(value: string): void {
    const resultName = validate(i18n.t('forms.result_name', 'Result Name'), value, [
      Alphanumeric,
      StartIsNonNumeric
    ]);
    this.setState({
      resultName,
      valid: this.state.valid && !hasErrors(resultName)
    });
  }

  private handleSave(): void {
    if (this.state.valid) {
      this.props.updateRouter(stateToNode(this.props.nodeSettings, this.state));
      this.props.onClose(false);
    }
  }

  private getButtons(): ButtonSet {
    return {
      primary: { name: i18n.t('buttons.ok', 'Ok'), onClick: this.handleSave },
      secondary: {
        name: i18n.t('buttons.cancel', 'Cancel'),
        onClick: () => this.props.onClose(true)
      }
    };
  }

  public renderEdit(): JSX.Element {
    const typeConfig = this.props.typeConfig;

    const advanced: Tab = {
      name: i18n.t('forms.advanced', 'Advanced'),
      body: (
        <CheckboxElement
          name={i18n.t('forms.all_destinations', 'All Destinations')}
          title={i18n.t('forms.all_destinations', 'All Destinations')}
          // todo labelClassName={styles.checkbox}
          checked={false} // todo checked={this.state.sendAll}
          description={i18n.t(
            'forms.all_destinations_description',
            "Send a message to all destinations known for this contact. If you aren't sure what this means, leave it unchecked."
          )}
          // todo onChange={this.handleSendAllUpdate}
        />
      )
      // todo checked: this.state.sendAll
    };

    const tabs = [advanced];

    return (
      <Dialog
        title={typeConfig.name}
        headerClass={typeConfig.type}
        buttons={this.getButtons()}
        tabs={tabs}
      >
        <TypeList __className="" initialType={typeConfig} onChange={this.props.onTypeChange} />
        <p>Enter the phone number to dial</p>
        <TextInputElement
          name="phone"
          placeholder={i18n.t('forms.phone_number', 'Phone Number')}
          showLabel={false}
          autocomplete={true}
          onChange={this.handlePhoneUpdated}
          entry={this.state.phone}
        />
        {createResultNameInput(this.state.resultName, this.handleUpdateResultName)}
        {renderIssues(this.props)}
      </Dialog>
    );
  }

  public render(): JSX.Element {
    return this.renderEdit();
  }
}
