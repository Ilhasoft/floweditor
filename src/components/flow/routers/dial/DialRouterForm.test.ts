import { RouterFormProps } from 'components/flow/props';
import DialRouterForm from 'components/flow/routers/dial/DialRouterForm';
import { Types } from 'config/interfaces';
import { composeComponentTestUtils, mock } from 'testUtils';
import { getRouterFormProps, createMatchRouter } from 'testUtils/assetCreators';
import * as utils from 'utils';
import { createUUID } from 'utils';

mock(utils, 'createUUID', utils.seededUUIDs());

const routerNode = createMatchRouter([]);
routerNode.ui = {
  position: { left: 0, top: 0 },
  type: Types.wait_for_dial
};

const { setup } = composeComponentTestUtils<RouterFormProps>(
  DialRouterForm,
  getRouterFormProps(routerNode)
);

describe(DialRouterForm.name, () => {
  it('should render', () => {
    const { wrapper } = setup(true);
    expect(wrapper).toMatchSnapshot();
  });

  describe('updates', () => {
    it('should save changes', () => {
      const { instance, props } = setup(true, {
        $merge: { updateRouter: jest.fn(), onClose: jest.fn() }
      });

      instance.handleUpdateResultName('Dial Result');
      instance.handlePhoneUpdated('@fields.supervisor');

      expect(instance.state).toMatchSnapshot();

      instance.handleSave();
      expect(props.onClose).toHaveBeenCalled();
      expect(props.updateRouter).toHaveBeenCalled();
      expect(props.updateRouter).toMatchCallSnapshot();
    });

    it('should cancel', () => {
      const { instance, props } = setup(true, {
        $merge: { updateRouter: jest.fn(), onClose: jest.fn() }
      });

      instance.handlePhoneUpdated('+1234567890');
      instance.getButtons().secondary.onClick();
      expect(props.onClose).toHaveBeenCalled();
      expect(props.updateRouter).not.toHaveBeenCalled();
    });
  });
});
