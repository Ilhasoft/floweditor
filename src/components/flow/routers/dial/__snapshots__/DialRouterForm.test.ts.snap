// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DialRouterForm should render 1`] = `
<Dialog
  buttons={
    Object {
      "primary": Object {
        "name": "Ok",
        "onClick": [Function],
      },
      "secondary": Object {
        "name": "Cancel",
        "onClick": [Function],
      },
    }
  }
  headerClass="split_by_dial"
  title="Redirect Call"
>
  <TypeList
    __className=""
    initialType={
      Object {
        "description": "Redirect call to another number",
        "form": [Function],
        "name": "Redirect Call",
        "type": "split_by_dial",
        "visibility": Array [
          "voice",
        ],
      }
    }
    onChange={[MockFunction]}
  />
  <p>
    If the expression...
  </p>
  <TextInputElement
    autocomplete={true}
    entry={
      Object {
        "value": undefined,
      }
    }
    name="Phone Number"
    onChange={[Function]}
    showLabel={false}
  />
  <OptionalTextInput
    helpText={
      <Trans
        i18nKey="forms.result_name_help"
        values={
          Object {
            "resultFormat": "@results",
          }
        }
      >
        By naming the result, you can reference it later using [[resultFormat]]
      </Trans>
    }
    maxLength={64}
    name="Result Name"
    onChange={[Function]}
    toggleText="Save as.."
    value={
      Object {
        "value": "",
      }
    }
  />
</Dialog>
`;

exports[`DialRouterForm updates should save changes 1`] = `
Object {
  "phone": Object {
    "validationFailures": Array [],
    "value": "@fields.supervisor",
  },
  "resultName": Object {
    "validationFailures": Array [],
    "value": "Dial Result",
  },
  "valid": true,
}
`;

exports[`DialRouterForm updates should save changes 2`] = `
Array [
  Object {
    "inboundConnections": Object {},
    "node": Object {
      "actions": Array [],
      "exits": Array [
        Object {
          "destination_uuid": null,
          "uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
        },
        Object {
          "destination_uuid": null,
          "uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
        },
        Object {
          "destination_uuid": null,
          "uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
        },
        Object {
          "destination_uuid": null,
          "uuid": "132de855-4042-4dc1-a18f-cc2e6a8f790a",
        },
      ],
      "router": Object {
        "cases": Array [
          Object {
            "arguments": Array [
              "answered",
            ],
            "category_uuid": "2dc85899-0577-4f1b-a620-f12094e34b5e",
            "type": "has_only_text",
            "uuid": "e0bc3776-792f-49d2-b4be-48be2cf8159a",
          },
          Object {
            "arguments": Array [
              "no_answer",
            ],
            "category_uuid": "42ebf1d0-a43e-4ba6-9577-c863b20c88c4",
            "type": "has_only_text",
            "uuid": "6c22884b-0e35-4206-982e-18320691eda9",
          },
          Object {
            "arguments": Array [
              "busy",
            ],
            "category_uuid": "70ac6ea6-803a-4c33-81c7-dc26803c313f",
            "type": "has_only_text",
            "uuid": "a62e8284-49f8-4292-adbc-d5f6607f85c5",
          },
        ],
        "categories": Array [
          Object {
            "exit_uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
            "name": "Answered",
            "uuid": "2dc85899-0577-4f1b-a620-f12094e34b5e",
          },
          Object {
            "exit_uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
            "name": "No Answer",
            "uuid": "42ebf1d0-a43e-4ba6-9577-c863b20c88c4",
          },
          Object {
            "exit_uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
            "name": "Busy",
            "uuid": "70ac6ea6-803a-4c33-81c7-dc26803c313f",
          },
          Object {
            "exit_uuid": "132de855-4042-4dc1-a18f-cc2e6a8f790a",
            "name": "Failed",
            "uuid": "e67ce051-f3e4-4c42-807b-e393df5fa382",
          },
        ],
        "default_category_uuid": "e67ce051-f3e4-4c42-807b-e393df5fa382",
        "operand": "@(default(resume.dial.status, \\"\\"))",
        "result_name": "Dial Result",
        "type": "switch",
        "wait": Object {
          "phone": "@fields.supervisor",
          "type": "dial",
        },
      },
      "uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
    },
    "ui": Object {
      "config": Object {},
      "position": null,
      "type": "split_by_dial",
    },
  },
]
`;
